diff --git a/lib/Makefile.am b/lib/Makefile.am
index 3c1dca7..1c2ae99 100644
--- a/lib/Makefile.am
+++ b/lib/Makefile.am
@@ -3,7 +3,7 @@
 AM_CFLAGS  = -Wall -I${top_srcdir}/include -D_GNU_SOURCE
 
 lib_LTLIBRARIES = \
-	libnl.la libnl-genl.la libnl-nf.la libnl-route.la
+	libnl.la libnl-genl.la libnl-route.la libnl-nf.la
 
 libnl_la_LDFLAGS = -version-info 2:0:0
 libnl_la_SOURCES = \
diff --git a/src/Makefile.am b/src/Makefile.am
index 0ffc26f..da2a9f5 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -20,67 +20,67 @@ noinst_PROGRAMS = \
 rtnl_utils_c = rtnl-utils.c utils.c
 
 genl_ctrl_list_SOURCES = genl-ctrl-list.c ctrl-utils.c utils.c
-genl_ctrl_list_LDADD = -lnl-genl
+genl_ctrl_list_LDADD = -lnl-genl -lnl
 
 nf_ct_list_SOURCES = nf-ct-list.c ct-utils.c utils.c
-nf_ct_list_LDADD = -lnl-nf
+nf_ct_list_LDADD = -lnl-nf -lnl-route -lnl
 nf_log_SOURCES = nf-log.c log-utils.c ${rtnl_utils_c}
-nf_log_LDADD = -lnl-nf
+nf_log_LDADD = -lnl-nf -lnl-route -lnl
 nf_queue_SOURCES = nf-queue.c queue-utils.c ${rtnl_utils_c}
-nf_queue_LDADD = -lnl-nf
+nf_queue_LDADD = -lnl-nf -lnl-route -lnl
 
 addr_c = addr-utils.c ${rtnl_utils_c}
 nl_addr_add_SOURCES = nl-addr-add.c ${addr_c}
-nl_addr_add_LDADD = -lnl-route
+nl_addr_add_LDADD = -lnl-route -lnl
 nl_addr_delete_SOURCES = nl-addr-delete.c ${addr_c}
-nl_addr_delete_LDADD = -lnl-route
+nl_addr_delete_LDADD = -lnl-route -lnl
 nl_addr_list_SOURCES = nl-addr-list.c ${addr_c}
-nl_addr_list_LDADD = -lnl-route
+nl_addr_list_LDADD = -lnl-route -lnl
 
 link_c = link-utils.c ${rtnl_utils_c}
 nl_link_list_SOURCES = nl-link-list.c ${link_c}
-nl_link_list_LDADD = -lnl-route
+nl_link_list_LDADD = -lnl-route -lnl
 nl_link_set_SOURCES = nl-link-set.c ${link_c}
-nl_link_set_LDADD = -lnl-route
+nl_link_set_LDADD = -lnl-route -lnl
 nl_link_stats_SOURCES = nl-link-stats.c ${link_c}
-nl_link_stats_LDADD = -lnl-route
+nl_link_stats_LDADD = -lnl-route -lnl
 nl_link_ifindex2name_SOURCES = nl-link-ifindex2name.c ${rtnl_utils_c}
-nl_link_ifindex2name_LDADD = -lnl-route
+nl_link_ifindex2name_LDADD = -lnl-route -lnl
 nl_link_name2ifindex_SOURCES = nl-link-name2ifindex.c ${rtnl_utils_c}
-nl_link_name2ifindex_LDADD = -lnl-route
+nl_link_name2ifindex_LDADD = -lnl-route -lnl
 
 nl_monitor_SOURCES = nl-monitor.c ${rtnl_utils_c}
-nl_monitor_LDADD = -lnl-route
+nl_monitor_LDADD = -lnl-route -lnl
 
 neigh_c = neigh-utils.c ${rtnl_utils_c}
 nl_neigh_add_SOURCES = nl-neigh-add.c ${neigh_c}
-nl_neigh_add_LDADD = -lnl-route
+nl_neigh_add_LDADD = -lnl-route -lnl
 nl_neigh_delete_SOURCES = nl-neigh-delete.c ${neigh_c}
-nl_neigh_delete_LDADD = -lnl-route
+nl_neigh_delete_LDADD = -lnl-route -lnl
 nl_neigh_list_SOURCES = nl-neigh-list.c ${neigh_c}
-nl_neigh_list_LDADD = -lnl-route
+nl_neigh_list_LDADD = -lnl-route -lnl
 
 nl_neightbl_list_SOURCES = nl-neightbl-list.c ${rtnl_utils_c}
-nl_neightbl_list_LDADD = -lnl-route
+nl_neightbl_list_LDADD = -lnl-route -lnl
 
 qdisc_c = qdisc-utils.c ${rtnl_utils_c}
 nl_qdisc_delete_SOURCES = nl-qdisc-delete.c ${qdisc_c}
-nl_qdisc_delete_LDADD = -lnl-route
+nl_qdisc_delete_LDADD = -lnl-route -lnl
 nl_qdisc_list_SOURCES = nl-qdisc-list.c ${qdisc_c}
-nl_qdisc_list_LDADD = -lnl-route
+nl_qdisc_list_LDADD = -lnl-route -lnl
 
 route_c = route-utils.c ${rtnl_utils_c}
 nl_route_add_SOURCES = nl-route-add.c ${route_c}
-nl_route_add_LDADD = -lnl-route
+nl_route_add_LDADD = -lnl-route -lnl
 nl_route_delete_SOURCES = nl-route-delete.c ${route_c}
-nl_route_delete_LDADD = -lnl-route
+nl_route_delete_LDADD = -lnl-route -lnl
 nl_route_get_SOURCES = nl-route-get.c ${route_c}
-nl_route_get_LDADD = -lnl-route
+nl_route_get_LDADD = -lnl-route -lnl
 nl_route_list_SOURCES = nl-route-list.c ${route_c}
-nl_route_list_LDADD = -lnl-route
+nl_route_list_LDADD = -lnl-route -lnl
 
 nl_rule_list_SOURCES = nl-rule-list.c rule-utils.c ${rtnl_utils_c}
-nl_rule_list_LDADD = -lnl-route
+nl_rule_list_LDADD = -lnl-route -lnl
 
 nl_tctree_list_SOURCES = nl-tctree-list.c ${rtnl_utils_c}
-nl_tctree_list_LDADD = -lnl-route
+nl_tctree_list_LDADD = -lnl-route -lnl
